react/features/gifs/middleware.js

```diff
@@ -0,0 +1,59 @@
+import { MiddlewareRegistry } from '../base/redux';
+
+import { ADD_GIF_FOR_PARTICIPANT, HIDE_GIF_FOR_PARTICIPANT, SHOW_GIF_FOR_PARTICIPANT } from './actionTypes';
+import { removeGif } from './actions';
+import { GIF_DEFAULT_TIMEOUT, GIF_HOVER_TIMEOUT } from './constants';
+import { getGifForParticipant } from './functions';
+
+/**
+ * Middleware which intercepts Gifs actions to handle changes to the
+ * visibility timeout of the Gifs.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(store => next => action => {
+    const { dispatch, getState } = store;
+    const state = getState();
+
+    switch (action.type) {
+    case ADD_GIF_FOR_PARTICIPANT: {
+        const id = action.participantId;
+        const { gif } = state['features/base/config'];
+
+        _clearGifTimeout(state, id);
+        const timeoutID = setTimeout(() => dispatch(removeGif(id)), gif?.tileTime || GIF_DEFAULT_TIMEOUT);
+
+        action.timeoutID = timeoutID;
+        break;
+    }
+    case SHOW_GIF_FOR_PARTICIPANT: {
+        const id = action.participantId;
+
+        _clearGifTimeout(state, id);
+        break;
+    }
+    case HIDE_GIF_FOR_PARTICIPANT: {
+        const id = action.participantId;
+        const timeoutID = setTimeout(() => dispatch(removeGif(id)), GIF_HOVER_TIMEOUT);
```

Comment:
Nit: if I hover and unhover it can happen that gif is hidden quicker than if I don't hover. eg i hover after 1 second and unhover immediately, it will show for 4 seconds while if I don t it will show for 5 with defaults. Is this expected?

Reply:
Not anymore. The requirements changed so I'll update this as well